apiserver/uniter/subordinaterelationwatcher.go

```diff
@@ -0,0 +1,166 @@
+// newSubordinateRelationsWatcher creates a watcher that will notify
+// about relation lifecycle events for subordinateApp, but filtered to
+// be relevant to a unit deployed to a container with the
+// principalName app. Global relations will be included, but only
+// container-scoped relations for the principal application will be
+// emitted - other container-scoped relations will be filtered out.
+func newSubordinateRelationsWatcher(backend *state.State, subordinateApp *state.Application, principalName string) (
+	state.StringsWatcher, error,
+) {
+	w := &subRelationsWatcher{
+		backend:       backend,
+		app:           subordinateApp,
+		principalName: principalName,
+		relations:     make(map[string]bool),
+		out:           make(chan []string),
+	}
+	err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+	})
+	return w, errors.Trace(err)
+}
+
+func (w *subRelationsWatcher) loop() error {
+	defer close(w.out)
+	relationsw := w.app.WatchRelations()
+	if err := w.catacomb.Add(relationsw); err != nil {
+		return errors.Trace(err)
+	}
+	var (
+		sentInitial bool
+		outputData  []string
+		out         chan []string
+
+		currentRelations = set.NewStrings()
+	)
+	for {
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+		case out <- outputData:
+			sentInitial = true
+			currentRelations = set.NewStrings()
+			outputData = nil
+			out = nil
+		case newRelations, ok := <-relationsw.Changes():
+			if !ok {
+				return watcher.EnsureErr(relationsw)
+			}
+			for _, relation := range newRelations {
+				if currentRelations.Contains(relation) {
+					continue
+				}
+				shouldSend, err := w.shouldSend(relation)
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if shouldSend {
+					currentRelations.Add(relation)
+					outputData = append(outputData, relation)
+				}
+			}
+			if !sentInitial || len(outputData) > 0 {
+				out = w.out
+			}
+		}
+	}
+}
+
+func (w *subRelationsWatcher) shouldSend(key string) (bool, error) {
+	if shouldSend, found := w.relations[key]; found {
+		return shouldSend, nil
+	}
+	result, err := w.shouldSendCheck(key)
+	if err == nil {
+		w.relations[key] = result
+	}
+	return result, errors.Trace(err)
+}
+
+func (w *subRelationsWatcher) shouldSendCheck(key string) (bool, error) {
+	rel, err := w.backend.KeyRelation(key)
+	if errors.IsNotFound(err) {
+		// We never saw it, and it's already gone away, so we can drop it.
+		logger.Debugf("couldn't find unknown relation %q", key)
+		return false, nil
+	} else if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	thisEnd, err := rel.Endpoint(w.app.Name())
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	if thisEnd.Scope == charm.ScopeGlobal {
+		return true, nil
+	}
+
+	// Only allow container relations if the other end is our principal.
+	otherEnds, err := rel.RelatedEndpoints(w.app.Name())
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	// Not really sure why RelatedEndpoints returns a slice -
```

Comment:
Ohh, I think I missed that it could be a peer role - in that case `counterpartRole` would return peer and both endpoints might be returned. But there'd never be more than 1 result in a provides/requires relation.

Comment removed, anyway.